AssertionError when .sass-lint.yml or package.json missing

I would expect sass-lint to work by just running:

    $ sass-lint

Or at least it should provide a helpful error message describing which required options I'm missing. Instead it crashes:

```
assert.js:89
  throw new assert.AssertionError({
  ^
AssertionError: missing path
    at Module.require (module.js:364:3)
    at require (module.js:385:17)
    at module.exports (/usr/local/lib/node_modules/sass-lint/lib/config.js:73:12)
    at Function.sassLint.getConfig (/usr/local/lib/node_modules/sass-lint/index.js:25:10)
    at Function.sassLint.lintFiles (/usr/local/lib/node_modules/sass-lint/index.js:166:18)
    at detectPattern (/usr/local/lib/node_modules/sass-lint/bin/sass-lint.js:16:18)
    at Object.<anonymous> (/usr/local/lib/node_modules/sass-lint/bin/sass-lint.js:87:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
```

So I follow the manual and create `.sass-lint.yml` and run it again:

    $ sass-lint

Again it crashes with the same stack trace as before.

So I go and look at this `config.js` file at line 73. Turns out it tries to read in `package.json`. Apparently that's a hidden feature implemented in #47, but documented nowhere.

So I create an empty `package.json` file consisting of just `{}` to please sass-lint.

Now finally I'm able to run it:

    $ sass-lint

But I see no errors, it appears to lint just fine, although I purposefully created a very buggy SCSS file. OK, I'll try to introduce a syntax error. I'll run the linter again, and it crashes:

```
/usr/local/lib/node_modules/sass-lint/index.js:250
    throw new Error(errorCount.count + ' errors were detected in \n- ' + errorCount.files.join('\n- '));
    ^

Error: 1 errors were detected in 
- test.scss
    at Function.sassLint.failOnError (/usr/local/lib/node_modules/sass-lint/index.js:250:11)
    at detectPattern (/usr/local/lib/node_modules/sass-lint/bin/sass-lint.js:27:10)
    at Object.<anonymous> (/usr/local/lib/node_modules/sass-lint/bin/sass-lint.js:87:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:134:18)
    at node.js:961:3
```

That error doesn't really help me either. I search through the issues in this repository, and find #500, from where I finally learn that I need to run:

    $ sass-lint --no-exit --verbose

The [CLI docs](https://github.com/sasstools/sass-lint/tree/develop/docs/cli) don't explain any of this. There's no way I would have guessed that I need to use `--verbose` option. Issue #500 mentions that there are plans to improve this in 2.0, but I really suggest documenting the current behavior.

Using sass-lint 1.5.1

--- a/index.js
+++ b/index.js
@@ -164,7 +164,6 @@ sassLint.lintFiles = function (files, options, configPath) {
   }
   else {
     files = this.getConfig(options, configPath).files;
-
     if (typeof files === 'string') {
       files = glob.sync(files);
     }

--- a/lib/config.js
+++ b/lib/config.js
@@ -14,29 +14,27 @@ var loadDefaults = function loadDefaults () {
 
 var findFile = function findFile (configPath, filename) {
   var HOME = process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE,
-      dirname,
-      parentDirname;
+      dirname = null,
+      parentDirname = null;
 
   configPath = configPath || path.join(process.cwd(), filename);
 
-  if (fs.existsSync(configPath)) {
+  if (configPath && fs.existsSync(configPath)) {
+    dirname = path.dirname(configPath);
+    parentDirname = path.dirname(dirname);
     return fs.realpathSync(configPath);
   }
 
-  dirname = path.dirname(configPath);
-  parentDirname = path.dirname(dirname);
-
-  if (dirname === HOME || dirname === parentDirname) {
+  if (dirname === null || dirname === HOME || dirname === parentDirname) {
     return null;
   }
-
   configPath = path.join(parentDirname, filename);
 
   return findFile(configPath, filename);
 };
 
 module.exports = function (options, configPath) {
-  var meta,
+  var meta = null,
       metaPath,
       configMerge = false,
       configMergeExists = false,
@@ -70,10 +68,11 @@ module.exports = function (options, configPath) {
 
   if (!configPath) {
     metaPath = findFile(false, 'package.json');
-    meta = require(metaPath);
+    if (metaPath) {
+      meta = require(metaPath);
+    }
 
-    if (meta.sasslintConfig) {
-
+    if (meta && meta.sasslintConfig) {
       configPath = path.resolve(path.dirname(metaPath), meta.sasslintConfig);
     }
     else {
@@ -90,7 +89,6 @@ module.exports = function (options, configPath) {
       config.rules = config.rules ? config.rules : {};
     }
   }
-
   // check to see if user config contains an options property and whether property has a property called merge-default-rules
   configMergeExists = (config.options && typeof config.options['merge-default-rules'] !== 'undefined');
 

--- a/tests/cli.js
+++ b/tests/cli.js
@@ -2,14 +2,14 @@ var assert = require('assert'),
     should = require('should'),
     fs = require('fs-extra'),
     path = require('path'),
-    childProcess = require('child_process');
+    exec = require('child_process').exec;
 
 
 describe('cli', function () {
   it('should return help instructions', function (done) {
     var command = 'sass-lint -h';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
       if (err) {
         return done(err);
       }
@@ -23,7 +23,7 @@ describe('cli', function () {
   it('should return a version', function (done) {
     var command = 'sass-lint -V';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
       if (err) {
         return done(err);
       }
@@ -37,7 +37,7 @@ describe('cli', function () {
   it('CLI format option should output JSON', function (done) {
     var command = 'sass-lint -c tests/yml/.stylish-output.yml tests/sass/cli.scss --verbose --format json';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       if (err) {
         return done(err);
@@ -58,7 +58,7 @@ describe('cli', function () {
     var command = 'sass-lint -c tests/yml/.stylish-output.yml tests/sass/cli.scss --verbose --format json --output tests/cli-output.json',
         outputFile = path.resolve(process.cwd(), 'tests/cli-output.json');
 
-    childProcess.exec(command, function (err) {
+    exec(command, function (err) {
 
       if (err) {
         return done(err);
@@ -81,7 +81,7 @@ describe('cli', function () {
     var command = 'sass-lint -c tests/yml/.stylish-output.yml tests/sass/cli.scss --verbose --format json --output tests/cli-output.json',
         outputFile = path.resolve(process.cwd(), 'tests/cli-output.json');
 
-    childProcess.exec(command, function (err) {
+    exec(command, function (err) {
 
       if (err) {
         return done(err);
@@ -114,7 +114,7 @@ describe('cli', function () {
     var command = 'sass-lint -c tests/yml/.stylish-output.yml tests/sass/cli.scss --verbose --format JSON --output tests/cli-output.json',
         outputFile = path.resolve(process.cwd(), 'tests/cli-output.json');
 
-    childProcess.exec(command, function (err) {
+    exec(command, function (err) {
 
       if (err) {
         return done(err);
@@ -148,7 +148,7 @@ describe('cli', function () {
   it('should return JSON from a custom config', function (done) {
     var command = 'sass-lint -c tests/yml/.color-keyword-errors.yml tests/sass/cli.scss --verbose';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       if (err) {
         return done(err);
@@ -170,7 +170,7 @@ describe('cli', function () {
   it('output should return no errors/warnings', function (done) {
     var command = 'sass-lint -c tests/yml/.json-lint.yml tests/sass/cli.scss --verbose';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       var result = 0;
 
@@ -192,7 +192,7 @@ describe('cli', function () {
   it('should return a warning', function (done) {
     var command = 'sass-lint -c tests/yml/.color-keyword-errors.yml tests/sass/cli.scss --verbose';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       var result = '';
 
@@ -222,7 +222,7 @@ describe('cli', function () {
     var command = 'sass-lint -c tests/yml/.stylish-errors.yml tests/sass/cli.scss --verbose',
         expectedOutputLength = 155;
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       if (err) {
         return done(err);
@@ -244,7 +244,7 @@ describe('cli', function () {
 
     var command = 'sass-lint -i \'**/*.s+(a|c)ss\'';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       if (err) {
         return done(err);
@@ -267,7 +267,7 @@ describe('cli', function () {
 
     var command = 'sass-lint -i \'**/*.scss, **/*.sass \'';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       if (err) {
         return done(err);
@@ -290,7 +290,7 @@ describe('cli', function () {
 
     var command = 'sass-lint --syntax scss tests/sass/cli.txt --verbose';
 
-    childProcess.exec(command, function (err, stdout) {
+    exec(command, function (err, stdout) {
 
       var result = 0;
 
@@ -311,12 +311,24 @@ describe('cli', function () {
   it('should exit with exit code 1 when quiet', function (done) {
     var command = 'sass-lint -c tests/yml/.error-output.yml tests/sass/cli-error.scss --verbose --no-exit';
 
-    childProcess.exec(command, function (err) {
+    exec(command, function (err) {
       if (err.code === 1) {
         return done();
       }
 
       return done(new Error('Error code not 1'));
+    });
+  });
+
+  it('should not exit with an error if no config is specified', function (done) {
+    var command = 'sass-lint tests/sass/cli-clean.scss --verbose --no-exit';
+
+    exec(command, function (err) {
+      if (!err) {
+        return done();
+      }
+
+      return done(new Error('Exited with error code 1'));
     });
   });
 
@@ -326,7 +338,7 @@ describe('cli', function () {
   it('parse errors should report as a lint error', function (done) {
     var command = 'sass-lint --config tests/yml/.stylish-output.yml tests/sass/parse.scss --verbose --no-exit --format json';
 
-    childProcess.exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
+    exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
       var result = JSON.parse(stdout)[0];
 
       assert.equal(1, result.errorCount);
@@ -337,7 +349,7 @@ describe('cli', function () {
   it('parse errors should report as severity 2', function (done) {
     var command = 'sass-lint --config tests/yml/.stylish-output.yml tests/sass/parse.scss --verbose --no-exit --format json';
 
-    childProcess.exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
+    exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
       var result = JSON.parse(stdout)[0],
           messages = result.messages[0],
           severity = 2;
@@ -350,7 +362,7 @@ describe('cli', function () {
   it('parse errors should report the correct message', function (done) {
     var command = 'sass-lint --config tests/yml/.stylish-output.yml tests/sass/parse.scss --verbose --no-exit --format json';
 
-    childProcess.exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
+    exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
       var result = JSON.parse(stdout)[0],
           message = result.messages[0].message,
           expected = 'Please check validity of the block starting from line #5';
@@ -363,7 +375,7 @@ describe('cli', function () {
   it('parse errors rule Id should be \'Fatal\'', function (done) {
     var command = 'sass-lint --config tests/yml/.stylish-output.yml tests/sass/parse.scss --verbose --no-exit --format json';
 
-    childProcess.exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
+    exec(command, function (err, stdout) { // eslint-disable-line handle-callback-err
       var result = JSON.parse(stdout)[0],
           messages = result.messages[0],
           ruleId = 'Fatal';

--- a/tests/sass/cli-clean.scss
+++ b/tests/sass/cli-clean.scss
@@ -1,0 +1,5 @@
+$red: #f00;
+
+.cli {
+  color: $red;
+}

